### ISSUE TITLE
problem with class ElementCount

### ISSUE THREAD

**Issue by pmqs** [avuserow/perl6-binary-structured#1]
Got the following error installing the module with panda.
```

t/030-inline.t ........... ok
Type check failed for return value; expected ElementCount but got Int (0)
  in method pull-elements at /media/paul/Linux-Shared/base/perl6/git/Archive-SimpleZip/.panda-work/1483022772_1/lib/Binary/Structured.pm6 (Binary::Structured) line 275
  in method <anon> at t/040-array.t line 16
  in block  at /media/paul/Linux-Shared/base/perl6/git/Archive-SimpleZip/.panda-work/1483022772_1/lib/Binary/Structured.pm6 (Binary::Structured) line 423
  in method parse at /media/paul/Linux-Shared/base/perl6/git/Archive-SimpleZip/.panda-work/1483022772_1/lib/Binary/Structured.pm6 (Binary::Structured) line 382
  in any call_with_capture at gen/moar/Metamodel.nqp line 3901
  in method parse at /media/paul/Linux-Shared/base/perl6/git/Archive-SimpleZip/.panda-work/1483022772_1/lib/Binary/Structured.pm6 (Binary::Structured) line 362
  in block <unit> at t/040-array.t line 2

```
Running the bleeding edge perl6.
```
$ perl6 -v
This is Rakudo version 2016.12-131-g19df358 built on MoarVM version 2016.12-23-g3c5253c
implementing Perl 6.c
```.

**Comment by kybr** [comment]
TL;DR ... I've fixed this error on my system and submitted a pull request to this repo. This project has some issues due to changes in Perl6.

- - -

Long story...

ElementCount is defined as...

```Perl6
class ElementCount is Int {}
```

Before 2016.10, this would create a new type identical to Int, but with the name ElementCount, but on/after that date, the definition above leads to different behaviour:

```Perl6
class ElementCount is Int {}
my $e = ElementCount.new(5);
say $e.WHAT;
# says ElementCount before 2016.10
# says Int on/after 2016.10
```

So, the type information is not retained on/after 2016.10 and the code in this project depends on the old behaviour. To fix this, I've defined ElementCount differently:

```Perl6
class ElementCount does Real {
  has $.n;
  multi method new(Real $r) {
    self.bless(:data($r));
  }
  multi method new(Int $i) {
    self.bless(:data($i));
  }
  multi sub infix:<+>(ElementCount $a, ElementCount $b) {
    ElementCount.new(n => $a.n + $b.n);
  }
  method Bridge() { $.n.Bridge }
  method perl() { "ElementCount.new(n => $.n)"; }
  method isNaN() { $.n.isNaN; }
}
```

This allows ElementCount to behave like an Int, but have a different type. I am certain that this could be done better, but it seems to work.

**Comment by avuserow** [comment]
Thanks for your contribution.
